### ISSUE TITLE
Add authentication to the demo api and PWA

### ISSUE THREAD

**Issue by artandor** [api-platform/demo#218]
**Description**  
I think it would be extremely valuable to add an example of authentication through JWT in the PWA, since it's something present in most projects. Maybe a refresh token could be fun too. Considering how easy it is to add new entities to the front thanks to the generator, it seems that the greatest improvement that can be made today is to make the auth easy aswell.

**Comment by vincentchalamon** [comment]
Hi @artandor,

There is already an authentication on the PWA on the admin: https://demo.api-platform.com/admin

**Comment by artandor** [comment]
It's handled by reactadmin, which makes things way easier, but yeah that's true.

**Comment by metaclass-nl** [comment]
[MetacClass Tutorial Api Platform](https://github.com/metaclass-nl/tutorial-api-platform) chapter 7 is about Authentication (JWT) and its [react branch](https://github.com/metaclass-nl/tutorial-api-platform/tree/chapter7-react) addresses the generated react client, but not in the PWA container because [the docs](https://api-platform.com/docs/client-generator/react/#react-generator) do not explain how to generate the react client there, so for now the the old client container is used. The tutorial can be extended with more branches for other clients but so far no one has come forward to do so.

**Comment by vincentchalamon** [comment]
@artandor You can try https://next-auth.js.org. I just implemented it on a next.js app, it's quite easy to implement even for a non-frontend developer :-)

**Comment by artandor** [comment]
I've read the next-auth a few time, maybe I misunderstood it but I thought next auth was creating an authentication service on the server, not connect to an existing one.So the auth flow with next app is client > next server (next Auth) > API platform Auth, which returns a token to next Auth and done ?Anyway thanks for your helpLe 3 août 2021 19:05, Vincent ***@***.***> a écrit :
@artandor You can try https://next-auth.js.org. I just implemented it on a next.js app, it's quite easy to implement even for a non-frontend developer :-)

**Comment by vincentchalamon** [comment]
I've created a `pages/api/auth/[...nextauth].ts` file in my PWA to configure next-auth with a custom provider: https://gist.github.com/vincentchalamon/ef71a8903b5dc9b19ce95ff1a2840d31.
I also needed to configure the NEXTAUTH_URL environment variable. Then, I could use `useSession` and/or `getSession` in my PWA pages to retrieve the current user. Apparently, next-auth stores what it wants as local-storage/session-storage.

To make it work with my example, I've created an AuthenticationSuccessEventSubscriber in my API to update the response in case of authentication success, to add the current user data, and the expiration of its token.
